### ISSUE TITLE
How to define an object type in Gatsby?

### ISSUE THREAD

**Issue by wangyi7099** [gatsbyjs/gatsby#16030]
## Summary
Hi! I want to pass a object named **webConfig**  to context in `createPages` action like this:
```js
        return createPage({
          path,
          component: docsTemplate,
          context: {
            webConfig
          },
        });
```
But the property and value in webConfig is dynamic and gatsby reports a warning : 

---
 warn There are conflicting field types in your data.

If you have explicitly defined a type for those fields, you can safely ignore this warning message.
Otherwise, Gatsby will omit those fields from the GraphQL schema.

If you know all field types in advance, the best strategy is to explicitly define them with the `createTypes` action, and skip inference with the `@dontInfer`
directive.

SitePage.context.webConfig.themeConfig.locales./.sidebar./guide/:
 - type: [string,object]
   value: [ 'aaa', { title: 'Guide', collapsable: false, children: [Array] }, ... ]
   source: 7374ebf2-d961-52ee-92a2-c25e7cb387a9
---

So I specify `webConfig` type in `createSchemaCustomization.js` 

```js
  createTypes(`
 type SitePage implements Node @infer {
      context: SitePageContext
    }

    type SitePageContext  {
      webConfig: JSON
    }
`); 
```

But the warning still show. I have no idea with how to deal with it..   Any suggestion? Thanks!

## Relevant information
none

### Environment (if relevant)

none
### File contents (if changed)

`gatsby-config.js`: N/A 
`package.json`: N/A 
`gatsby-node.js`: N/A 
`gatsby-browser.js`: N/A 
`gatsby-ssr.js`: N/A

**Comment by stefanprobst** [comment]
When type inference is enabled, the warning message will unfortunately always show -- this is what the first line `If you have explicitly defined a type for those fields, you can safely ignore this warning message.` refers to.

To totally get rid of if you could try to fully define the `SitePage` type and add the `dontInfer` directive -- probably all you need is `path`:
```graphql
type SitePage implements Node @dontInfer {
  path: String
}
```

Note that you don't need to have the `context` field available in the schema unless you want to query for that field, i.e. do something like `query { sitePage { context { webConfig } } }`.
The context value will be available as a variable in your page query and as `pageContext` prop in the component regardless of it being in the schema or not.

**Comment by wangyi7099** [comment]
Thanks for the replay@stefanprobst. I have tried 
```js
type SitePage implements Node @dontInfer {
      path: String
    }
```
 Unfortunately, the warning still shows.. :(

**Comment by stefanprobst** [comment]
You are right! This is actually a bug, should be fixed in #16110 - thanks for reporting!

**Comment by wangyi7099** [comment]
Hi Looks good now. But could you pelase have a look at this issue https://github.com/gatsbyjs/gatsby/issues/15544 ? I have created a repo for it.
